modules/javafx.controls/src/main/java/javafx/scene/control/skin/MenuBarSkin.java

```diff
@@ -389,62 +276,141 @@ public MenuBarSkin(final MenuBar control) {
+                                        }
+                                        ev.consume();
+                                        return;
+                                    }
+                                    if (isRTL) {
+                                        moveToMenu(Direction.PREVIOUS, true);
+                                    } else {
+                                        moveToMenu(Direction.NEXT, true);
+                                    }
+                                }
+                                ev.consume();
+                                break;
+                            }
+                            case DOWN:
+                            //case SPACE:
+                            //case ENTER:
+                                // RT-18859: Doing nothing for space and enter
+                                if (control.getScene().getWindow().isFocused()) {
+                                    if (focusedMenuIndex != -1) {
+                                        Menu menuToOpen = getSkinnable().getMenus().get(focusedMenuIndex);
+                                        showMenu(menuToOpen, true);
+                                        ev.consume();
+                                    }
+                                }
+                                break;
+                            case ESCAPE:
+                                unSelectMenus();
+                                ev.consume();
+                                break;
+                        default:
+                            break;
+                        }
+                    }
+                });
+
+                // When we click else where in the scene - menu selection should be cleared.
+                sceneListenerHelper.addEventFilter(scene, MouseEvent.MOUSE_CLICKED, (ev) -> {
+                    Bounds containerScreenBounds = container.localToScreen(container.getLayoutBounds());
+                    if ((containerScreenBounds == null) || !containerScreenBounds.contains(ev.getScreenX(), ev.getScreenY())) {
+                        unSelectMenus();
+                    }
+                });
+
+                // When the parent window looses focus - menu selection should be cleared
+                sceneListenerHelper.addChangeListener(scene.windowProperty(), true, (sr, oldw, w) -> {
+                    if (windowFocusHelper != null) {
+                        windowFocusHelper.disconnect();
+                        windowFocusHelper = null;
+                    }
+
+                    if (w != null) {
+                        windowFocusHelper = sceneListenerHelper.addChangeListener(w.focusedProperty(), true, (s, p, focused) -> {
+                            if (!focused) {
+                                unSelectMenus();
+                            }
+                        });
+                    }
+                });
```

Comment:
The goal seems to be to subscribe to changes on scene->window->focused.

This can be done with `flatMap` / `orElse`.

```suggestion
                // When the parent window looses focus - menu selection should be cleared
                sceneListenerHelper.addChangeListener(
                    scene.windowProperty()
                         .flatMap(Window::focusedProperty)
                         .orElse(false), 
                    true, 
                    focused -> {
                        if (!focused) {
                            unSelectMenus();
                        }
                    }
                );
```

You could even do this at the top level I think:

    lh.addChangeListener(
        control.sceneProperty()
            .flatMap(Scene::windowProperty)
            .flatMap(Window::focusedProperty)
            .orElse(false), 
        true, 
        focused -> {
            if (!focused) {
                unSelectMenus();
            }
        }
    );

Also available is to make use of `Subscription`:

        Subscription sub = Subscription.subscribe(scene.windowProperty().flatMap(Window::focusedProperty).orElse(false), focused -> {
              if (!focused) {
                  unSelectMenus();
              }
        });

The subscription can then be integrated with `ListenerHelper` again (or even more directly if it accepted `Subscription`):

        lh.addDisconnectable(sub::unsubscribe);

Reply:
Good suggestions.  
It is basically the existing code - I'd prefer to keep it as is, since re-writing it might introduce regression.